ngraph/core/reference/include/ngraph/runtime/reference/adaptive_max_pool.hpp

```diff
@@ -0,0 +1,99 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <cfenv>
+#include <cmath>
+#include <numeric>
+#include <stdexcept>
+#include <vector>
+
+#include "ngraph/axis_vector.hpp"
+#include "ngraph/coordinate_transform.hpp"
+#include "ngraph/shape.hpp"
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            template <typename T>
+            void adaptive_max_pool(const T* arg,
+                                   T* out,
+                                   int64_t* selected_indices,
+                                   const Shape& arg_shape,
+                                   const Shape& out_shape)
+            {
+                // At the outermost level we will walk over every output coordinate O.
+                CoordinateTransform output_transform(out_shape);
```

Comment:
This implementation uses `CoordinateTransform` which was identified as having low performance. Our team did **a lot** of refactoring to remove its usage across all of the reference implementations and yet here it is introduced again in new implementation. Why is that? Do you have good reason to that? If no good reason then please re-write this implementation and calculate all of the indexes manually. 

CC: @ilyachur

Reply:
This implementation is based on an already existing pool reference implementation, only difference is window size calculation. I didn't know we consider `CoordinateTransform` as deprecated or banned for use.

Reply:
@ilyachur can we mark `CoordinateTransform` as deprecated to avoid such cases in the future?

Reply:
Yes we can

Reply:
@mvafin Could you don't use `CoordinateTransform`? This class has a significant impact on performance.